File: src/main/java/no/nav/syfo/consumer/ws/BehandleJournalConsumer.java
package no.nav.syfo.consumer.ws;

import lombok.extern.slf4j.Slf4j;
import no.nav.syfo.controller.PDFRestController;
import no.nav.syfo.domain.Soknad;
import no.nav.syfo.domain.dto.PDFTemplate;
import no.nav.syfo.domain.dto.Soknadstype;
import no.nav.tjeneste.virksomhet.behandlejournal.v2.BehandleJournalV2;
import no.nav.tjeneste.virksomhet.behandlejournal.v2.informasjon.behandlejournal.*;
import no.nav.tjeneste.virksomhet.behandlejournal.v2.informasjon.journalfoerinngaaendehenvendelse.WSDokumentinfoRelasjon;
import no.nav.tjeneste.virksomhet.behandlejournal.v2.informasjon.journalfoerinngaaendehenvendelse.WSJournalfoertDokumentInfo;
import no.nav.tjeneste.virksomhet.behandlejournal.v2.informasjon.journalfoerinngaaendehenvendelse.WSJournalpost;
import no.nav.tjeneste.virksomhet.behandlejournal.v2.meldinger.WSJournalfoerInngaaendeHenvendelseRequest;
import org.springframework.stereotype.Component;

import javax.inject.Inject;
import java.time.LocalDateTime;

import static no.nav.syfo.domain.dto.PDFTemplate.*;
import static no.nav.syfo.util.DatoUtil.norskDato;

@Component
@Slf4j
public class BehandleJournalConsumer {

    private BehandleJournalV2 behandleJournalV2;
    private PersonConsumer personConsumer;
    private PDFRestController pdfRestController;

    private static final String GOSYS = "FS22";

    @Inject
    public BehandleJournalConsumer(
            BehandleJournalV2 behandleJournalV2,
            PersonConsumer personConsumer,
            PDFRestController pdfRestController) {
        this.behandleJournalV2 = behandleJournalV2;
        this.personConsumer = personConsumer;
        this.pdfRestController = pdfRestController;
    }

    public String opprettJournalpost(Soknad soknad, String saksId) {
        byte[] pdf;

        try {
            pdf = pdfRestController.getPDF(soknad, hentPDFTemplateEtterSoknadstype(soknad.getSoknadstype()));
        } catch (RuntimeException e) {
            String feilmelding = "Kunne ikke generere PDF for søknad med id: " + soknad.getSoknadsId() + " og saks id: " + saksId;
            log.error(feilmelding, e);
            throw new RuntimeException(feilmelding, e);
        }

        return journalforSoknad(soknad, saksId, pdf);
    }

    private String journalforSoknad(Soknad soknad, String saksId, byte[] pdf) {
        try {
            return behandleJournalV2.journalfoerInngaaendeHenvendelse(
                    new WSJournalfoerInngaaendeHenvendelseRequest()
                            .withApplikasjonsID("SYFOGSAK")
                            .withJournalpost(new WSJournalpost()
                                    .withDokumentDato(LocalDateTime.now())
                                    .withJournalfoerendeEnhetREF(GOSYS)
                                    .withKanal(new WSKommunikasjonskanaler().withValue("NAV_NO"))
                                    .withSignatur(new WSSignatur().withSignert(true))
                                    .withArkivtema(new WSArkivtemaer().withValue("SYK"))
                                    .withForBruker(new WSPerson().withIdent(new WSNorskIdent().withIdent(soknad.getFnr())))
                                    .withOpprettetAvNavn("Syfogsak")
                                    .withInnhold(getJournalPostInnholdNavn(soknad.getSoknadstype()))
                                    .withEksternPart(new WSEksternPart()
                                            .withNavn(personConsumer.finnBrukerPersonnavnByFnr(soknad.getFnr()))
                                            .withEksternAktoer(new WSPerson().withIdent(new WSNorskIdent().withIdent(soknad.getFnr()))))
                                    .withGjelderSak(new WSSak().withSaksId(saksId).withFagsystemkode(GOSYS))
                                    .withMottattDato(LocalDateTime.now())
                                    .withDokumentinfoRelasjon(
                                            new WSDokumentinfoRelasjon()
                                                    .withTillknyttetJournalpostSomKode("HOVEDDOKUMENT")
                                                    .withJournalfoertDokument(new WSJournalfoertDokumentInfo()
                                                            .withBegrensetPartsInnsyn(false)
                                                            .withDokumentType(new WSDokumenttyper().withValue("ES"))
                                                            .withSensitivitet(true)
                                                            .withTittel(getJornalfoertDokumentTittel(soknad))
                                                            .withKategorikode("ES")
                                                            .withBeskriverInnhold(
                                                                    new WSStrukturertInnhold()
                                                                            .withFilnavn(getWSStruktureltInnholdFilnavn(soknad))
                                                                            .withFiltype(new WSArkivfiltyper().withValue("PDF"))
                                                                            .withInnhold(pdf)
                                                                            .withVariantformat(new WSVariantformater().withValue("ARKIV"))
                                                            ))
                                    ))
            ).getJournalpostId();
        } catch (RuntimeException e) {
            String feilmelding = "Kunne ikke behandle journalpost for søknad med id " + soknad.getSoknadsId() + " og saks id: " + saksId;
            log.error(feilmelding, e);
            throw new RuntimeException(feilmelding, e);
        }
    }

    private String getJornalfoertDokumentTittel(Soknad soknad) {
        switch (soknad.getSoknadstype()) {
            case OPPHOLD_UTLAND:
                return "Søknad om å beholde sykepenger utenfor Norge";
            case SELVSTENDIGE_OG_FRILANSERE:
                return "Søknad om sykepenger fra Selvstendig/Frilanser for periode: " + soknad.getFom().format(norskDato) + " til " + soknad.getTom().format(norskDato);
            case ARBEIDSTAKERE:
                return "Søknad om sykepenger " + soknad.getFom().format(norskDato) + " - " + soknad.getTom().format(norskDato);
            default:
                throw new RuntimeException("Har ikke implementert journalført dokumenttittel for søknad av typen: " + soknad.getSoknadstype());
        }
    }

    private String getWSStruktureltInnholdFilnavn(Soknad soknad) {
        switch (soknad.getSoknadstype()) {
            case OPPHOLD_UTLAND:
                return "soknad-" + soknad.getInnsendtDato().format(norskDato);
            case SELVSTENDIGE_OG_FRILANSERE:
                return "Søknad om sykepenger fra Selvstendig/Frilanser for periode: " + soknad.getFom().format(norskDato) + " til " + soknad.getTom().format(norskDato);
            case ARBEIDSTAKERE:
                return "Søknad om sykepenger " + soknad.getFom().format(norskDato) + " - " + soknad.getTom().format(norskDato);
            default:
                throw new RuntimeException("Har ikke implementert strukturert innhold-filnavn for søknad av typen: " + soknad.getSoknadstype());
        }
    }

    private String getJournalPostInnholdNavn(Soknadstype soknadstype) {
        switch (soknadstype) {
            case OPPHOLD_UTLAND:
                return "Søknad om å beholde sykepenger utenfor Norge";
            case SELVSTENDIGE_OG_FRILANSERE:
            case ARBEIDSTAKERE:
                return "Søknad om sykepenger";
            default:
                throw new RuntimeException("Har ikke implementert strukturert innhold-filnavn for søknad av typen: " + soknadstype);
        }
    }

    private PDFTemplate hentPDFTemplateEtterSoknadstype(Soknadstype soknadstype) {
        switch (soknadstype) {
            case OPPHOLD_UTLAND:
                return SYKEPENGERUTLAND;
            case SELVSTENDIGE_OG_FRILANSERE:
                return SELVSTENDIGNAERINGSDRIVENDE;
            case ARBEIDSTAKERE:
                return ARBEIDSTAKERE;
            default:
                throw new RuntimeException("Har ikke implementert PDF-template for søknad av typen: " + soknadstype);
        }
    }
}

File: src/main/kotlin/no/nav/syfo/consumer/ws/BehandleJournalConsumer.kt
package no.nav.syfo.consumer.ws

import no.nav.syfo.consumer.pdf.PDFConsumer
import no.nav.syfo.domain.Soknad
import no.nav.syfo.domain.dto.PDFTemplate
import no.nav.syfo.domain.dto.PDFTemplate.SELVSTENDIGNAERINGSDRIVENDE
import no.nav.syfo.domain.dto.PDFTemplate.SYKEPENGERUTLAND
import no.nav.syfo.domain.dto.Soknadstype
import no.nav.syfo.domain.dto.Soknadstype.*
import no.nav.syfo.logger
import no.nav.syfo.util.DatoUtil.norskDato
import no.nav.tjeneste.virksomhet.behandlejournal.v2.BehandleJournalV2
import no.nav.tjeneste.virksomhet.behandlejournal.v2.informasjon.behandlejournal.*
import no.nav.tjeneste.virksomhet.behandlejournal.v2.informasjon.journalfoerinngaaendehenvendelse.DokumentinfoRelasjon
import no.nav.tjeneste.virksomhet.behandlejournal.v2.informasjon.journalfoerinngaaendehenvendelse.JournalfoertDokumentInfo
import no.nav.tjeneste.virksomhet.behandlejournal.v2.informasjon.journalfoerinngaaendehenvendelse.Journalpost
import no.nav.tjeneste.virksomhet.behandlejournal.v2.meldinger.JournalfoerInngaaendeHenvendelseRequest
import org.joda.time.DateTime
import org.springframework.stereotype.Component
import javax.inject.Inject

@Component
class BehandleJournalConsumer @Inject
constructor(
    private val behandleJournalV2: BehandleJournalV2,
    private val personConsumer: PersonConsumer,
    private val pdfConsumer: PDFConsumer
) {
    private val log = logger()

    fun opprettJournalpost(soknad: Soknad, saksId: String): String {
        val pdf: ByteArray?

        try {
            pdf = pdfConsumer.getPDF(soknad, hentPDFTemplateEtterSoknadstype(soknad.soknadstype))
        } catch (e: RuntimeException) {
            val feilmelding = "Kunne ikke generere PDF for søknad med id: ${soknad.soknadsId} og saks id: $saksId"
            log.error(feilmelding, e)
            throw RuntimeException(feilmelding, e)
        }

        return journalforSoknad(soknad, saksId, pdf)
    }

    private fun journalforSoknad(soknad: Soknad, saksId: String, pdf: ByteArray?): String {
        try {
            return behandleJournalV2.journalfoerInngaaendeHenvendelse(
                JournalfoerInngaaendeHenvendelseRequest()
                    .withApplikasjonsID("SYFOGSAK")
                    .withJournalpost(
                        Journalpost()
                            .withDokumentDato(DateTime.now())
                            .withJournalfoerendeEnhetREF(JOURNALFORENDE_ENHET)
                            .withKanal(Kommunikasjonskanaler().withValue("NAV_NO"))
                            .withSignatur(Signatur().withSignert(true))
                            .withArkivtema(Arkivtemaer().withValue("SYK"))
                            .withForBruker(Person().withIdent(NorskIdent().withIdent(soknad.fnr)))
                            .withOpprettetAvNavn("Syfogsak")
                            .withInnhold(getJournalPostInnholdNavn(soknad.soknadstype))
                            .withEksternPart(
                                EksternPart()
                                    .withNavn(personConsumer.finnBrukerPersonnavnByFnr(soknad.fnr!!))
                                    .withEksternAktoer(Person().withIdent(NorskIdent().withIdent(soknad.fnr)))
                            )
                            .withGjelderSak(Sak().withSaksId(saksId).withFagsystemkode(GOSYS))
                            .withMottattDato(DateTime.now())
                            .withDokumentinfoRelasjon(
                                DokumentinfoRelasjon()
                                    .withTillknyttetJournalpostSomKode("HOVEDDOKUMENT")
                                    .withJournalfoertDokument(
                                        JournalfoertDokumentInfo()
                                            .withBegrensetPartsInnsyn(false)
                                            .withDokumentType(Dokumenttyper().withValue(getBrevkode(soknad)))
                                            .withSensitivitet(true)
                                            .withTittel(getJornalfoertDokumentTittel(soknad))
                                            .withKategorikode("ES")
                                            .withBeskriverInnhold(
                                                StrukturertInnhold()
                                                    .withFilnavn(getStruktureltInnholdFilnavn(soknad))
                                                    .withFiltype(Arkivfiltyper().withValue("PDF"))
                                                    .withInnhold(pdf)
                                                    .withVariantformat(Variantformater().withValue("ARKIV"))
                                            )
                                    )
                            )
                    )
            ).journalpostId
        } catch (e: RuntimeException) {
            val feilmelding = "Kunne ikke behandle journalpost for søknad med id ${soknad.soknadsId} og saks id: $saksId"
            log.error(feilmelding, e)
            throw RuntimeException(feilmelding, e)
        }
    }

    private fun getBrevkode(soknad: Soknad): String {
        return when (soknad.soknadstype) {
            OPPHOLD_UTLAND -> "NAV 08-07.09"
            REISETILSKUDD -> "NAV 08-14.01"
            SELVSTENDIGE_OG_FRILANSERE, ARBEIDSTAKERE, ARBEIDSLEDIG, BEHANDLINGSDAGER, ANNET_ARBEIDSFORHOLD -> "NAV 08-07.04 D"
        }
    }

    private fun getJornalfoertDokumentTittel(soknad: Soknad): String {
        return when (soknad.soknadstype) {
            OPPHOLD_UTLAND -> "Søknad om å beholde sykepenger utenfor EØS"
            SELVSTENDIGE_OG_FRILANSERE -> "Søknad om sykepenger fra Selvstendig/Frilanser for periode: ${soknad.fom!!.format(norskDato)} til ${soknad.tom!!.format(norskDato)}"
            ARBEIDSTAKERE -> "Søknad om sykepenger ${soknad.fom!!.format(norskDato)} - ${soknad.tom!!.format(norskDato)}"
            ARBEIDSLEDIG -> "Søknad om sykepenger fra arbeidsledig for periode: ${soknad.fom!!.format(norskDato)} til ${soknad.tom!!.format(norskDato)}"
            BEHANDLINGSDAGER -> "Søknad om enkeltstående behandlingsdager fra ${soknad.arbeidssituasjon.toString().toLowerCase()} for periode: ${soknad.fom!!.format(norskDato)} til ${soknad.tom!!.format(norskDato)}"
            ANNET_ARBEIDSFORHOLD -> "Søknad om sykepenger med uavklart arbeidssituasjon fra ${soknad.fom!!.format(norskDato)} til ${soknad.tom!!.format(norskDato)}"
            REISETILSKUDD -> "Søknad om reisetilskudd for periode: ${soknad.fom!!.format(norskDato)} til ${soknad.tom!!.format(norskDato)}"
        }
    }

    private fun getStruktureltInnholdFilnavn(soknad: Soknad): String {
        return when (soknad.soknadstype) {
            OPPHOLD_UTLAND -> "soknad-${soknad.innsendtDato!!.format(norskDato)}"
            SELVSTENDIGE_OG_FRILANSERE -> "Søknad om sykepenger fra Selvstendig/Frilanser for periode: ${soknad.fom!!.format(norskDato)} til ${soknad.tom!!.format(norskDato)}"
            ARBEIDSTAKERE -> "Søknad om sykepenger ${soknad.fom!!.format(norskDato)} - ${soknad.tom!!.format(norskDato)}"
            ARBEIDSLEDIG -> "Søknad om sykepenger fra arbeidsledig for periode: ${soknad.fom!!.format(norskDato)} til ${soknad.tom!!.format(norskDato)}"
            BEHANDLINGSDAGER -> "Søknad om enkeltstående behandlingsdager fra ${soknad.arbeidssituasjon.toString().toLowerCase()} for periode: ${soknad.fom!!.format(norskDato)} til ${soknad.tom!!.format(norskDato)}"
            ANNET_ARBEIDSFORHOLD -> "Søknad om sykepenger med uavklart arbeidssituasjon fra ${soknad.fom!!.format(norskDato)} til ${soknad.tom!!.format(norskDato)}"
            REISETILSKUDD -> "Søknad om reisetilskudd for periode: ${soknad.fom!!.format(norskDato)} til ${soknad.tom!!.format(norskDato)}"
        }
    }

    private fun getJournalPostInnholdNavn(soknadstype: Soknadstype): String {
        return when (soknadstype) {
            OPPHOLD_UTLAND -> "Søknad om å beholde sykepenger utenfor EØS"
            SELVSTENDIGE_OG_FRILANSERE, ARBEIDSTAKERE, ARBEIDSLEDIG, ANNET_ARBEIDSFORHOLD -> "Søknad om sykepenger"
            REISETILSKUDD -> "Søknad om reisetilskudd"
            BEHANDLINGSDAGER -> "Søknad om enkeltstående behandlingsdager"
        }
    }

    private fun hentPDFTemplateEtterSoknadstype(soknadstype: Soknadstype): PDFTemplate {
        return when (soknadstype) {
            OPPHOLD_UTLAND -> SYKEPENGERUTLAND
            SELVSTENDIGE_OG_FRILANSERE -> SELVSTENDIGNAERINGSDRIVENDE
            ARBEIDSTAKERE -> PDFTemplate.ARBEIDSTAKERE
            ARBEIDSLEDIG -> PDFTemplate.ARBEIDSLEDIG
            BEHANDLINGSDAGER -> PDFTemplate.BEHANDLINGSDAGER
            ANNET_ARBEIDSFORHOLD -> PDFTemplate.ANNETARBEIDSFORHOLD
            REISETILSKUDD -> PDFTemplate.REISETILSKUDD
        }
    }

    companion object {
        private val GOSYS = "FS22"
        private val JOURNALFORENDE_ENHET = "9999"
    }
}
